Add more shuffle options for dungeon items

tanjo3 opened the pull request:
This PR adds more shuffle options for dungeon items.

Previously, the options were to either have _all_ dungeon items shuffled within their own dungeons or to shuffle _all_ dungeon items anywhere (key-lunacy). This PR first splits the dungeon items into three categories, each with its own shuffle option: small keys, Big Keys, and Dungeon Maps and Compasses.

The new shuffle options are:
- **Vanilla**: All items in that category will be in their vanilla (original) locations.
- **Start With**: The player starts with all items in that category in their inventory.
- **Own Dungeon**: All items in that category will be shuffled within progression locations in their original dungeon. This is the original default behavior of the randomizer.
- **Any Dungeon**: All items in that category will be shuffled within any progression location in any dungeon. If Required Bosses Mode is on, dungeon items of required bosses' dungeons will not appear in non-required bosses' dungeons.
- **Overworld**: All items in that category will be shuffled throughout progression overworld locations.
- **Anywhere**: All items in that category will be shuffled in any progression location. This is the original default behavior of key-lunacy. 

The UI of the randomizer has been updated to provide three dropdowns to allow the user to select these shuffle modes _independently_ of one another. The default is having all dungeon items shuffled to their own dungeons. The hint logic of the randomizer has also been updated to only hint at keys if they are shuffled in any dungeon, overworld, or anywhere. Similarly, only small keys shuffled in vanilla or own dungeon modes will not count as their own sphere.

The randomized starting gear list has also been updated to include dungeon keys, along with the maps and compasses. I also changed the list's behavior to filter out dungeon items when their shuffle mode was set to vanilla or start-with shuffle mode. Starting with the dungeon items is self-explanatory and is really a visual bugfix. I chose to remove the items in vanilla shuffle mode to avoid needing additional logic for which vanilla locations get randomized. For instance, with vanilla small keys and a DRC Small Key added to the starting gear, which DRC Small Key vanilla location gets randomized? It's not immediately obvious which is the best way to handle this, as there are multiple approaches.

Finally, I've added some tests for combinations of dungeon item shuffle modes. The first test method covers all combinations with default settings. The rest cover combinations of key shuffle modes in Required Bosses Mode with 4 (default) and 1 boss, and when dungeons are not progression. I did some more testing locally, but this is already 324 new tests, which should cover most cases.
**Edit:** Since tests were taking over 2 hours to run, I cut down the number of tests to 158 by removing maps and compasses as a parameter. I also maked most tests added in this PR as "thorough" so that they can still be run locally, but GH Actions will only run 36 of the newly-added tests.

<img src="https://github.com/user-attachments/assets/09bf6518-930f-48e7-a03c-39a3ff3825c3" width="600" />

tanjo3 committed:
Add options for dungeon item shuffle mode

tanjo3 committed:
Split dungeon item lists by type

tanjo3 committed:
Add logic for checking valid dungeon item placement

tanjo3 committed:
Add logic for determining the vanilla locations of items

tanjo3 committed:
Add logic to place dungeon items based on shuffle mode

tanjo3 committed:
Update UI to expose new shuffle options

Additionally, misc. cleanup of UI including renaming layouts, spacers, and groupboxes.

tanjo3 committed:
Update hints to account for new shuffle options

tanjo3 committed:
Add tests for dungeon item shuffle

wooferzfg commented:
I'm curious what the reasoning was for bosses, and minibosses when those entrances are randomized, to not have dungeon items. These were the commits that made these changes:
- https://github.com/LagoLunatic/wwrando/commit/efbec974e49a5e0fa05e1953dfcbbd238dcf9d22
- https://github.com/LagoLunatic/wwrando/commit/6271b412bf6f632bca5095a727b98ea447c921f0

tanjo3 committed:
Filter maps and compasses from starting gear in Vanilla and Start With modes

tanjo3 committed:
Add dungeon keys to randomized starting gear list

tanjo3 commented:
I updated the randomized starting gear list to include dungeon keys, along with the maps and compasses. I also changed the list's behavior to filter out dungeon items when their shuffle mode was set to Vanilla or Start With. Start With is self-explanatory and is really a visual bugfix. I chose to remove the items in Vanilla shuffle mode to avoid needing additional logic for which vanilla locations get randomized. For instance, with vanilla small keys and a DRC Small Key added to the starting gear, which DRC Small Key vanilla location gets randomized? It's not immediately obvious which is the best way to handle this, as there are multiple approaches.

tanjo3 committed:
Properly categorize small keys as progressive

tanjo3 committed:
Don't hint at dungeon items when their shuffle mode is set to vanilla

tanjo3 committed:
Add missing type hints

tanjo3 committed:
Fix indentation

tanjo3 committed:
Fix bad change to location hints

tanjo3 committed:
Fix tabstops

tanjo3 committed:
Fix DRC Small Keys not being encoded in the permalink properly

tanjo3 committed:
Fix inconsistency with how dungeon small keys are considered

tanjo3 committed:
Ensure option name matches enum value

tanjo3 committed:
Remove unused import

tanjo3 committed:
Remove reference to key-lunacy from comment

tanjo3 committed:
Loosen dungeon item placement when they can be shuffled outside of their own dungeon

tanjo3 committed:
Prevent ANY_DUNGEON items from being placed in unreachable dungeons

tanjo3 committed:
Retry randomization up to 10 times upon failure

tanjo3 committed:
Use specific exception for dungeon item placement failures

tanjo3 committed:
Minor comment fix

tanjo3 committed:
Avoid duplicating starting dungeon items

This can't happen through the UI, since it removes dungeon items from the starting gear list if its shuffle mode is set to START_WITH. However, it could still happen with a stale/edited settings.txt.

tanjo3 committed:
Remove starting keys from progress item count

tanjo3 committed:
Retry entire item forward fill, rather than just dungeon item fill

tanjo3 committed:
Only pre-place ANY_DUNGEON dungeon items when Required Bosses Mode is off

With a dungeons-only seed with ANY_DUNGEON small keys, pre-placing keys resulted in only the DRC First Room location to be sphere 0, resulting in generation failures. Pre-placement is only necessary to handle banned dungeons, so use natural forward fill otherwise.

tanjo3 committed:
Remove maps and compasses from tests and only run a subset through Actions

tanjo3 committed:
Merge remote-tracking branch 'origin/master' into keysanity-rework

tanjo3 committed:
Merge remote-tracking branch 'origin/master' into keysanity-rework

tanjo3 committed:
Collapse redundant variables

tanjo3 committed:
Narrow retry to typed placement errors

tanjo3 committed:
Fix misleading error label

tanjo3 committed:
Fix gclib version from merge
